File: packages/typescript-llm-conversation/src/llmSubmitShotgun.ts
import { AIClientLike, identifyLLMProvider } from './llmProviders.js';
import { llmSubmit, LLMSubmitOptions } from './llmSubmit.js';

/**
 * Implements the "shotgun" strategy for {@link llmSubmit}: sends `numBarrels`
 * parallel requests with identical inputs, then asks the model to examine all
 * responses and reconcile them into a single authoritative answer.
 *
 * This improves output quality for tasks where the model benefits from
 * exploring multiple reasoning paths simultaneously.
 *
 * @param messages - The conversation history to send to each worker.
 * @param aiClient - The LLM provider's client instance.
 * @param options - Submit options (the `shotgun` field is stripped before
 *   forwarding to avoid infinite recursion).
 * @param numBarrels - Number of parallel worker requests to fire.
 * @returns The reconciled response from the model.
 */
export const llmSubmitShotgun = async (
  messages: unknown[],
  aiClient: AIClientLike,
  options: LLMSubmitOptions,
  numBarrels: number
): Promise<
  string | Record<string, unknown> | unknown[] | number | boolean | null
> => {
  messages = JSON.parse(JSON.stringify(messages));
  options = JSON.parse(JSON.stringify(options));

  // Delete the shotgun option from the options passed to
  // llmSubmitShotgun to avoid infinite recursion.
  delete options.shotgun;

  if (numBarrels <= 1) {
    // No need for shotgun logic if only 1 barrel requested.
    return llmSubmit(messages, aiClient, options);
  }

  // Remove the shotgun option before passing to llmSubmit to avoid
  // infinite recursion!
  delete options.shotgun;

  const convoShotgun: unknown[][] = [];
  for (let i = 0; i < numBarrels; i += 1) {
    const convoBarrel = JSON.parse(JSON.stringify(messages));
    convoShotgun.push(convoBarrel);
  }
  const resultsRaw: unknown[] = await Promise.all(
    convoShotgun.map((convoBarrel) => llmSubmit(convoBarrel, aiClient, options))
  );
  const resultStrings = resultsRaw.map((result) => JSON.stringify(result));

  // Build the reconciliation conversation on top of a fresh copy of the original messages.
  const reconcileMessages: unknown[] = JSON.parse(JSON.stringify(messages));

  reconcileMessages.push({
    role: 'system',
    content: `
SYSTEM MESSAGE:
In order to produce better results, we submitted this request/question/command/etc.
to ${numBarrels} worker threads in parallel.
The system will now present each of their responses, wrapped in JSON.
The user or developer will not see these responses -- they are only for you, the assistant, 
to examine and reconcile. Think of them as brainstorming or scratchpads.
`,
  });
  resultStrings.forEach((resultString, index) => {
    reconcileMessages.push({
      role: 'system',
      content: `WORKER ${index + 1} RESPONSE:\n\n\n${resultString}`,
    });
  });

  reconcileMessages.push({
    role: 'system',
    content: `
Focus on the differences and discrepancies between the workers' responses. Where do they agree?
Where do they disagree? In the areas where they disagree, which worker's argument is most
consistent with the data you've been shown?

Remember, this is an adjudication, not a democracy -- you should carefully examine the data
presented in the conversation and use your best judgment to determine which worker is most
likely to be correct, even if they're in the minority. Evaluate their answers carefully against
the source data. If multiple workers produced different answers, then clearly there is something
subtle, deceptive, or misleading about the question or the data, and you should be especially
careful to scrutinize the workers' reasoning and the evidence they present for their answers.
At least one of them must be wrong; don't fall into the same trap he did.
`,
  });

  // This is a chain-of-thought ponderance. We specifically do not want a JSON
  // response here, because we want the model to be able to freely reason and
  // draw conclusions without being constrained by JSON syntax. The final answer
  // will be produced in the next step, where we will ask the model to produce
  // the same format as it was originally given (text or JSON).
  const sPonderReply = await llmSubmit(reconcileMessages, aiClient, {
    ...options,
    jsonResponse: undefined,
  });
  reconcileMessages.push({ role: 'assistant', content: sPonderReply });

  reconcileMessages.push({
    role: 'system',
    content: `
Having seen and reconciled the workers' responses, you are now ready to craft a proper reply to
the question/request/command/etc. This response that you craft now is the one that will be
presented to the user or developer -- it should not directly reference the workers' responses,
but should instead be a fully self-contained and complete answer that draws on the insights
you've gained from examining the workers' responses.
`,
  });

  return llmSubmit(reconcileMessages, aiClient, options);
};

File: packages/python-llm-conversation/src/mightydatainc_llm_conversation/llm_submit_shotgun.py
import json
from typing import Any


def llm_submit_shotgun(
    messages: list[Any],
    ai_client: Any,
    options: dict[str, Any],
    num_barrels: int,
):
    """Run shotgun parallel submissions and reconcile to one final answer."""
    from .llm_submit import llm_submit

    # Deep-copy messages so worker conversations cannot cross-mutate.
    messages = json.loads(json.dumps(messages))

    if num_barrels <= 1:
        return llm_submit(messages=messages, ai_client=ai_client, **options)

    # Remove shotgun before nested submits to prevent recursion.
    options_to_pass_to_workers = {**options}
    options_to_pass_to_workers.pop("shotgun", None)

    convo_shotgun: list[list[Any]] = []
    for _ in range(num_barrels):
        convo_barrel = json.loads(json.dumps(messages))
        convo_shotgun.append(convo_barrel)

    results_raw: list[Any] = [
        llm_submit(
            messages=convo_barrel,
            ai_client=ai_client,
            **options_to_pass_to_workers,
        )
        for convo_barrel in convo_shotgun
    ]
    result_strings = [json.dumps(result) for result in results_raw]

    reconcile_messages: list[Any] = json.loads(json.dumps(messages))

    reconcile_messages.append(
        {
            "role": "system",
            "content": f"""
SYSTEM MESSAGE:
In order to produce better results, we submitted this request/question/command/etc.
to {num_barrels} worker threads in parallel.
The system will now present each of their responses, wrapped in JSON.
The user or developer will not see these responses -- they are only for you, the assistant,
to examine and reconcile. Think of them as brainstorming or scratchpads.
""",
        }
    )

    for index, result_string in enumerate(result_strings):
        reconcile_messages.append(
            {
                "role": "system",
                "content": f"WORKER {index + 1} RESPONSE:\n\n\n{result_string}",
            }
        )

    reconcile_messages.append(
        {
            "role": "system",
            "content": """
Focus on the differences and discrepancies between the workers' responses. Where do they agree?
Where do they disagree? In the areas where they disagree, which worker's argument is most
consistent with the data you've been shown?

Remember, this is an adjudication, not a democracy -- you should carefully examine the data
presented in the conversation and use your best judgment to determine which worker is most
likely to be correct, even if they're in the minority. Evaluate their answers carefully against
the source data. If multiple workers produced different answers, then clearly there is something
subtle, deceptive, or misleading about the question or the data, and you should be especially
careful to scrutinize the workers' reasoning and the evidence they present for their answers.
At least one of them must be wrong; don't fall into the same trap he did.
""",
        }
    )

    s_ponder_reply = llm_submit(
        messages=reconcile_messages,
        ai_client=ai_client,
        **{**options_to_pass_to_workers, "json_response": None},
    )
    reconcile_messages.append({"role": "assistant", "content": s_ponder_reply})

    reconcile_messages.append(
        {
            "role": "system",
            "content": """
Having seen and reconciled the workers' responses, you are now ready to craft a proper reply to
the question/request/command/etc. This response that you craft now is the one that will be
presented to the user or developer -- it should not directly reference the workers' responses,
but should instead be a fully self-contained and complete answer that draws on the insights
you've gained from examining the workers' responses.
""",
        }
    )

    return llm_submit(
        messages=reconcile_messages,
        ai_client=ai_client,
        **options_to_pass_to_workers,
    )
